src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs

```diff
@@ -30197,6 +30208,119 @@ public static int TakeOutParam<T>(out T x)
+            Assert.True(x1.Type.IsErrorType());
+        }
+
+        [Fact, WorkItem(17321, "https://github.com/dotnet/roslyn/issues/17321")]
+        public void InferenceFailure_01()
+        {
+            string source =
+@"
+class H
+{
+    object M1() => M(M(1), x1);
+    static object M(object o1) => o1;
+    static void M(object o1, object o2) {}
+}
+";
+            var node0 = SyntaxFactory.ParseCompilationUnit(source);
+            var one = node0.DescendantNodes().OfType<LiteralExpressionSyntax>().Single();
+            var decl = SyntaxFactory.DeclarationExpression(
+                type: SyntaxFactory.IdentifierName(SyntaxFactory.Identifier("var")),
+                designation: SyntaxFactory.SingleVariableDesignation(SyntaxFactory.Identifier("x1")));
+            var node1 = node0.ReplaceNode(one, decl);
+            var tree = node1.SyntaxTree;
+            Assert.NotNull(tree);
+            var compilation = CreateCompilationWithMscorlib(new[] { tree });
+            compilation.VerifyDiagnostics(
+                // (4,24): error CS8185: A declaration is not allowed in this context.
+                //     object M1() => M(M(varx1), x1);
+                Diagnostic(ErrorCode.ERR_DeclarationExpressionNotPermitted, "varx1").WithLocation(4, 24)
+                );
+
+            var model = compilation.GetSemanticModel(tree);
+            var x1Decl = tree.GetRoot().DescendantNodes().OfType<DeclarationExpressionSyntax>()
+                    .Where(p => p.Identifier().ValueText == "x1").Single();
+            var x1Ref = GetReference(tree, "x1");
+            VerifyModelForDeclarationVarWithoutDataFlow(model, x1Decl, x1Ref);
+        }
+
+        [Fact, WorkItem(17321, "https://github.com/dotnet/roslyn/issues/17321")]
+        public void InferenceFailure_02()
+        {
+            string source =
+@"
+class H
+{
+    object M1() => 1;
+}
+";
+            var node0 = SyntaxFactory.ParseCompilationUnit(source);
+            var one = node0.DescendantNodes().OfType<LiteralExpressionSyntax>().Single();
+            var decl = SyntaxFactory.DeclarationExpression(
+                type: SyntaxFactory.IdentifierName(SyntaxFactory.Identifier("var").WithTrailingTrivia(SyntaxFactory.Space)),
+                designation: SyntaxFactory.SingleVariableDesignation(SyntaxFactory.Identifier("x1")));
+            var node1 = node0.ReplaceNode(one, decl);
+            var tree = node1.SyntaxTree;
+            Assert.NotNull(tree);
+            var compilation = CreateCompilationWithMscorlib(new[] { tree });
+            compilation.VerifyDiagnostics(
+                // (4,20): error CS8185: A declaration is not allowed in this context.
+                //     object M1() => var x1;
+                Diagnostic(ErrorCode.ERR_DeclarationExpressionNotPermitted, "var x1").WithLocation(4, 20)
+                );
+
+            // We do not currently have test helpers that support declaration expressions subject to a conversion.
```

Comment:
> The work you propose is outside the scope of the intended fix here.

I am not suggesting to change product behavior, i.e. the fix. I am suggesting that it is not adequately covered by the tests. How do we know if the fix is complete if you don't test it? It would take less than 5 minutes to parameterize existing helpers to create a specialized flavor that would handle your special case, or possibly to adjust the scenario so that it would be handled by the existing helpers. It would probably take less than 15 minutes to create a new brand new dedicated helper from scratch. Yet we are discussing this for more than 4 hours. I am not even asking to test new scenarios, just asking to do more thorough testing for the ones you already have.